packages/fuel-indexer-graphql/src/dynamic.rs

```diff
@@ -0,0 +1,457 @@
+        "UInt4",
+        "UInt8",
+        "UInt16",
+        "Timestamp",
+        "Color",
+        "ContractId",
+        "Salt",
+        "Json",
+        "MessageId",
+        "Charfield",
+        "Identity",
+        "Boolean",
+        "Blob",
+    ]);
+    static ref NUMERIC_SCALAR_TYPES: HashSet<&'static str> = HashSet::from([
+        "Int4",
+        "Int8",
+        "Int16",
+        "UInt4",
+        "UInt8",
+        "UInt16",
+        "Timestamp",
+    ]);
+    static ref STRING_SCALAR_TYPES: HashSet<&'static str> = HashSet::from([
+        "ID",
+        "Address",
+        "AssetId",
+        "Bytes4",
+        "Bytes8",
+        "Bytes32",
+        "Bytes64",
+        "Color",
+        "ContractId",
+        "Salt",
+        "Json",
+        "MessageId",
+        "Charfield",
+        "Identity",
+        "Blob",
+    ]);
+    static ref SORTABLE_SCALAR_TYPES: HashSet<&'static str> = HashSet::from([
+        "ID",
+        "Address",
+        "AssetId",
+        "Int4",
+        "Int8",
+        "Int16",
+        "UInt4",
+        "UInt8",
+        "UInt16",
+        "Timestamp",
+        "Color",
+        "ContractId",
+        "Salt",
+        "MessageId",
+        "Charfield",
+        "Identity",
+    ]);
+    static ref IGNORED_ENTITY_TYPES: HashSet<&'static str> =
+        HashSet::from(["IndexMetadataEntity", "QueryRoot"]);
+    static ref IGNORED_ENTITY_FIELD_TYPES: HashSet<&'static str> =
+        HashSet::from(["object"]);
+}
+
+/// Execute user query and return results.
+pub async fn execute_query(
+    dynamic_request: Request,
+    dynamic_schema: DynamicSchema,
+    user_query: String,
+    pool: IndexerConnectionPool,
+    schema: Schema,
+) -> Result<Value, GraphqlError> {
+    // Because the schema types from async-graphql expect each field to be resolved
+    // separately, it became untenable to use the .execute() method of the dynamic
+    // schema itself to resolve queries. Instead, we set it to only resolve
+    // introspection queries and then pass any non-introspection queries to our
+    // custom query resolver.
+    match dynamic_request.operation_name.as_deref() {
+        Some("IntrospectionQuery") | Some("introspectionquery") => {
+            let introspection_results = dynamic_schema.execute(dynamic_request).await;
+            let data = match introspection_results.data.into_json() {
+                Ok(json) => json,
+                Err(_) => {
+                    return Err(GraphqlError::DynamicQueryError(
+                        "Could not parse introspection response".to_string(),
+                    ))
+                }
+            };
+
+            Ok(serde_json::json!({ "data": data }))
+        }
+        Some(_) | None => {
+            let query =
+                GraphqlQueryBuilder::new(&schema, user_query.as_str())?.build()?;
+
+            let queries = query.as_sql(&schema, pool.database_type())?.join(";\n");
+            let mut conn = match pool.acquire().await {
```

Comment:
Can we unwrap with `?` here?

Reply:
I've chosen not to use a `?` here as we would need to pull in `sqlx` as a dependency for `fuel-indexer-graphql`; instead, I've added additional information to `GraphQlError`.